### ISSUE TITLE
Introduce UI for range-diff command

### ISSUE THREAD

**Issue by crypto-rsa** [gitextensions/gitextensions#8448]
## Feature description
Git 2.19 added a new feature [range-diff](https://git-scm.com/docs/git-range-diff) which is very useful for checking for example conflict resolutions in a rebased branch. It would be awesome to have some UI for it. It could be launched in a similar manner like the regular diff window, i.e. selecting first commit using _Compare - Select as BASE to compare (Ctrl + L)_ and then selecting the other commit.
The UI itself could show the output of the command `git range-diff {first}...{second}` as a list of commits which are either unique in one of the revisions or same in both or differing. There should be options to hide each "category" of commits. For differing commits, there should be an option to show the original commit from either branch or their "combined" diff as reported by the `range-diff` command.

## Environment

- Git Extensions 33.33.33
- Build 4992b04e6e70ed491cf333365668be1fb4d51125
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4220.0
- DPI 96dpi (no scaling)

**Comment by gerhardol** [comment]
GE 3.4 has something like this for the file contents. If you select commits in two branches, you get the diff to the base too.
There has been discussions about comparing contents in branches too, to see the diff for a range on two different branches (like click A,B  on one branch and C,D on another and see if the contents match.)

The documentation for range-diff mentions that the command is currently considered to be only human readable, so parsing can be difficult and have to be changed from one release to another.

The range info could be displayed as an extra group, just as the common diff to base.

**Comment by gerhardol** [comment]
I propose this is added automatically in the Diff tab whenever there is a common base, just as the specific diffs and common diff is shown now. (there will have to be some special handling of the diffs, similar to the handling in the GitHub PRs are presented.

A lot of groups though... Is an option to hide the range diff needed? (in addition Show differences for all parents, also in the context menu).
@mstv any comments?

**Comment by gerhardol** [comment]
A poc presenting the complete range-diff as a file
Obviously a new icon is needed, maybe special sorting too.
The original colors should be preserved if possible, but maybe I should finish #7825 first (for that PR a lot of other refactoring was requested though).

![image](https://user-images.githubusercontent.com/6248932/93015707-d76a9f80-f5bb-11ea-9b7a-f89801999de0.png)

**Comment by RussKie** [comment]
Should "RangeDiff" live in its own group?

**Comment by gerhardol** [comment]
> Should "RangeDiff" live in its own group?

It is maybe good to separate this presentation.
I will probably make a WIP PR for this in days/weeks.

**Comment by gerhardol** [comment]
> > Should "RangeDiff" live in its own group?
> 
> It is maybe good to separate this presentation.
> I will probably make a WIP PR for this in days/weeks.

As diff can include artificial diff and range diff cannot, the presentation must be a separate group, the contents is not the same. The raw contents from range diff is a first stage and also a fallback for parsing contents. A problem with parsing is that the full range diff need to be parsed both when finding the groups and when presenting the patch. Probably ok.

**Comment by gerhardol** [comment]
A draft PR, but before the actual change is reviewed, it need to be discussed.
The feature is of great help when reviewing changes with rebases.
The default printout has some nice coloring, that I did not get the fileviewer to display. Did not try too hard, tried to parse the git-range-diff outputs. The output is hard to parse, I will need to look at the Git code to get something mostly acceptable. Still much better than the raw output.

The output is however intended to be "human readable" and parsing will fail occasionally. The command take some time too. If you range-diff two commits and many commits are involved, a diff could require a minute (at home with minimal viruses and comparing with 10 year old GE branch).

Should the output be displayed by default?
I believe so, even if advanced. No option if not required (but see below)

The functionality requires the the commits compared have a common merge base (that is not the commits themselves).
Should the functionality be limited to selecting 3 or four commits?
(I do not think so, it will be a hidden feature otherwise.)

Should there be a max commit difference to calculate range-diff, to avoid "hanging" for big diffs?
I am a little hesitant, what hangs one PC will be too low for another.

Parsing will occasionally fail. I still want to try parsing, it works in most situations but we do not want to believe that GE has bugs.
Is it sufficient to add parse-leftovers to other 'files' to make it possible to see the contents or should this parsing be skipped?
If the parsing is buggy, should the raw output be displayed too?
(This is how it is done in the draft PR)

The parsed "files" are patches but without line numbers etc so not quite fully qualified patches (like the GitHub pull request patches are fake too). Options like ignore spaces and 'window' could have an effect but will require that the range-diff command is cached less effectively and that parsing has to be done several times.
Are options required here?
(If OK to skip, the toolbar should be customized too.)

"file" icons could be customized to show new files etc, or special icon (currently diff icon) is displayed for them

More question to come...
@mstv I know you have opinions, do you have time?

**Comment by crypto-rsa** [comment]
I took a quick look and I think it is very useful even in the most rudimentary form (i.e. showing the raw command output). Especially when combined with the normal diff - that shows _what_ changes were made but the range-diff also shows _which commit_ was each change made in. Also shows changes in commit messages, very useful! 👍

**Comment by gerhardol** [comment]
> I took a quick look and I think it is very useful even in the most rudimentary form (i.e. showing the raw command output). 

It sure is useful and few commits and minor changes is perfectly visible and possible to interpret only with raw output - but Git dual color is better.

With many commits and or major changes, the raw output is hard to follow, and thousands of commit differences is no point also for parsed output. So parsed output is a good addition, but if it seems buggy to the user, should we hide it?
Is it better to not offer functionality to the user if they can blame GE even if it is mostly useful?
(I have not studied the Git code closely enough to find if the output seem to be stable enough for at least a specific version.)

**Comment by mstv** [comment]
Looks very promising to me! Perfect timing: I'll begin with the integration of a vendor delivery tomorrow.

Unfortunately, the regarding git commands only appear in the command log very seldom. (Oh no, I didn't want to restart with a new bug report.)

**Comment by gerhardol** [comment]
Looking at the output some more and the Git code, I believe that it will not be possible to have a reasonable safe full parsing of the patch for a commit pair.
The range-diff code looks reasonable, but the output varies a lot. Parsing ca be mostly fine, but you occasionally need to look at the original raw output (that is not always correct either).
Often it is more important to find what commits that differ, rather than the details by file.

The proposal is therefore to parse the different commit pairs and present the raw output, like below.
If Git colors can be retrieved, that is even better.
The file viewer shows colors assuming that this is a normal patch, not quite OK.

The parsing is reasonable safe and the overview is better than the raw output.
Another enhancement would be to navigate to the revisions in the group, to more easily see details.
The number of lines with context and diff white space can be supported without too mush problems too. (But the Git command will have to run for every option change and output parsing be done every time a file is selected - should be OK).

![image](https://user-images.githubusercontent.com/6248932/94205818-5ddc8680-fec4-11ea-984e-32798e02d713.png)

**Comment by mstv** [comment]
You have asked, Gerhard ;-)

- The display of diffs has done too much on the UI thread already before this extension.
  Windows moves an unresponsive app instance to the end of the list of instances in the taskbar - a great nuisance with many open sorted GE windows.
  So I am afraid, the loading must be done on request or really cancelable and with low prio - or else in a dedicated window.
- The limit of 100 is too high. It happens too often that GE does not respond for a minute.
- If the range-diff contains only unmatched commits or long ranges of them, it is not worth to display (all of) them.
- Yes, git range-diff does not work meaningful with vendor branches, which have an unrelated history initially.
- Yes, even the git output is truncated occasionally in an inconsistent manner.
- Range-diff is another strong argument for switching to a TreeView.

**Comment by gerhardol** [comment]
> So I am afraid, the loading must be done on request or really cancelable and with low prio - or else in a dedicated window.

The processing when selecting commits could be done async or changed to expanding the groups, but it is not like that now.
But in the current implementation, all groups and "files" must be added when selecting the commits.
This would limit the presentation to one group with one file (so git-range-diff only executes when selecting the file.)
This limits the usefulness for me, harder to see what commands are changed.

What if the presentation above with two ranges selected only? (three or four clicks)
Would user detect the functionality at all?

Or present two virtual files; Summary and full?

The presentation of the raw output in the fileviewer can be improved, but to lit Git do the coloring would be simpler.

> The limit of 100 is too high. It happens too often that GE does not respond for a minute.

That is not final, set a little to high right now.
I personally work with patchsets with maybe 30 commits (in each branch), which is quite close to the limit.
The limit could be separate for two selected and four selected too.
With many selected, the list is more interested than the details.

**Comment by gerhardol** [comment]
IThe draft PR is mostly implemented. There can be improvements to the highlighting, but mostly feature complete. (A few TBD like Copy Patch includes leading space +/-)
The datastructures in FileViewer need to be cleaned up.

Navigation next/previous patch goes to next commit-compare
The commit-compare headers should maybe have at least a different shade than patch headers (there is no other user color)
The dual color in the git output is a little difficult to recreate.
(It could be possible to convert the git terminal output colors to file viewer too.)

![image](https://user-images.githubusercontent.com/6248932/94352583-3f62c080-0067-11eb-99d1-857212dfa8ed.png)

**Comment by crypto-rsa** [comment]
Thank you very much, @gerhardol! The feature made my code reviews much easier to do 👍
